lib/experimental/block-bindings/sources/pattern.php

```diff
@@ -0,0 +1,23 @@
+<?php
+/**
+ * Add the metadata source to the block bindings API.
+ *
+ * @package gutenberg
+ */
+
+if ( function_exists( 'register_block_bindings_source' ) ) {
+	$pattern_source_callback = function ( $source_attrs, $block_content, $block, $block_instance ) {
```

Comment:
The params for `apply` are probably the one thing in this PR that need to be carefully considered for the future (in terms of the back compat problems it might introduce). 

- `$source_attrs` - I'm not sure how this one works, it seems to pass an `attributes` property of the binding source's config, but it's unused in this PR (`$post_meta_source_callback` does try to use it, but it's not set).
- `$block_content` - It seems to be unusued, and I'm wondering what potential future usage there might be for this? I think it's also already available via `$block_instance->inner_html`, so perhaps not necessary as a param.
- `$block` / `$block_instance` - $block represents the parsed block object, which is also available using `$block_instance->parsed_block`, so I think it could be removed as a param.

I think it could at a minimum be reduced to two params.

Reply:
Regarding this, I passed those params because we are using the [render_block_{$this->name} hook](https://developer.wordpress.org/reference/hooks/render_block_this-name/) to process the bindings: [link](https://github.com/WordPress/gutenberg/pull/57249/files#diff-be8a354ce9c01e85c0fdf04c75d76880cdc6c81992db7a86a7a0515dc11a9833R169). As that filter is using those params, I decided to pass them in case they are needed and other sources want to use them.

Having said that, I agree it's worth analyzing if they are really needed and try to simplify the `apply`. To ensure we are aligned, you propose using `block_instance` for everything, right? For example, this would mean that the pattern and the post meta sources would be:

**Pattern**
```PHP
// Before
$pattern_source_callback = function ( $source_attrs, $block_content, $block, $block_instance ) {
   ...
}

// After
$pattern_source_callback = function ( $block_instance ) {
   ...
}
```

**Post meta**

```PHP
// Before
$post_meta_source_callback = function ( $source_attrs ) {
   ...
}

// After
$post_meta_source_callback = function ( $block_instance ) {
   $source_attrs = $block_instance->parsed_block['attrs']['metadata']['bindings'];
   ...
}
```

Is that what you had in mind?

For the sources we have created so far in the different experiments, in all of them we are just using the `source_attrs` aside from the `pattern` where we are using the `$block_instance`.

Reply:
My impression is that we should refactor the way bindings get registered first, and it will drive changes in how these filters work. In the WordPress core, we won’t use filters as far as I can tell from a  quick assessment so that might evolve into a bit different shape.

Reply:
I actually updated it to use just the values that are actually in use for now. We can change it back if we feel that will spark better discussion.